Feature: new assertOutgoingRequests command

felipevolpone opened the pull request:
## Proposed Changes
This PR implements a new command to assert outgoing requests from the app. This will be used to check if the app does some network calls for certain APIs/backend.
The `mock-server` will be used to fetch and match the events stored in Maestro Cloud.

This is the command definition:
```yaml
- assertOutgoingRequest:
    url: ".*api.company.com\/[^\/]+\/login"
    headersPresent:
      - "cache-control"
    headersAndValues:
      - "Content-Type": "*application/json*"
      - "Authorization": "Bearer .*"
    httpMethodIs: "GET"
    requestBodyContains: '"foo": "bar"'
```

Important to say that this command will only work if the user have set the Maestro SDK properly (https://maestro.mobile.dev/advanced/experimental/maestro-mock-server/getting-started#setup)

The implementation for Android and iOS relies on getting the sessionId (and UUID). For Android, the maestro-sdk generates the random UUID; for iOS the driver generates and the SDK fetches from there. With the sessionId, maestro can call the Maestro Cloud backend and filter the events and make the assert logic.

I know this is a very long PR and I'm sorry for that, but if you follow the commit timeline it will make it easier to review it.

dmitry-zaitsev commented on maestro-android/src/androidTest/java/dev/mobile/maestro/MaestroDriverService.kt:
```diff
@@ -287,6 +301,29 @@ class Service(
+            .url("http://localhost:7008/session")
+            .build()
+
+        val sessionInfo = httpClient
```
This call might throw an error, you need to propagate it properly to `responseObserver`

dmitry-zaitsev commented on maestro-android/src/androidTest/java/dev/mobile/maestro/MaestroDriverService.kt:
```diff
@@ -287,6 +301,29 @@ class Service(
+    ) {
+        val sessionReq = Request.Builder()
+            .get()
+            .url("http://localhost:7008/session")
```
This service is running in the same process, what is the point of using HTTP to communicate with it? You can just access memory directly

dmitry-zaitsev replied:
Specifically, you are looking for `SdkService` class

amanjeetsingh150 commented on maestro-android/src/androidTest/java/dev/mobile/maestro/MaestroDriverService.kt:
```diff
@@ -287,6 +287,14 @@ class Service(
+        request: MaestroAndroid.SessionRequest,
+        responseObserver: StreamObserver<MaestroAndroid.SessionResponse>
+    ) {
+        responseObserver.onNext(sessionResponse { id = SdkService.sessionInfo.sessionId })
```
Can session id be null?

felipevolpone replied:
No, the sessionId is the first thing created when the users uses the SDK

amanjeetsingh150 replied:
👍

amanjeetsingh150 commented on maestro-client/src/main/java/maestro/OutgoingRequest.kt:
```diff
@@ -0,0 +1,43 @@
+            eventsFilteredByHeadersAndValues.filter { e -> e.bodyAsString?.contains(requestBody) == true }
+        } ?: eventsFilteredByHeadersAndValues
+
+        println("from ${events.size} events, ${eventsMatching.size} match the url ${rules.url}")
```
nit: Capitalization `F`. If this to be shown to user

amanjeetsingh150 commented on maestro-client/src/main/java/maestro/drivers/IOSDriver.kt:
```diff
@@ -494,6 +495,10 @@ class IOSDriver(
     }
 
+    override fun fetchSessionId(): UUID {
+        return iosDevice.fetchSessionInfo().expect {  }
```
This will fail if fetchSessionInfo errors. Is this intended?

felipevolpone replied:
Yes

amanjeetsingh150 commented on maestro-orchestra-models/src/main/java/maestro/orchestra/Commands.kt:
```diff
@@ -748,6 +748,22 @@ data class TravelCommand(
+    }
+
+    override fun evaluateScripts(jsEngine: JsEngine): Command {
+        return this
```
Assertions can have JS resolutions from runScript output I would recommend using evaluate script in case of string.

amanjeetsingh150 commented on maestro-ios-xctest-runner/maestro-driver-iosUITests/Routes/Handlers/SessionInfoHandler.swift:
```diff
@@ -0,0 +1,33 @@
+
+@MainActor
+struct SessionInfoHandler: HTTPHandler {
+
```
I see changes in xctest but the updated binaries are not here?

felipevolpone replied:
How can I do that?

berikv replied:
(cd maestro; ./maestro-ios-xctest-runner/build-maestro-ios-runner.sh)

felipevolpone replied:
@berikv this command fails for me with long output. I'll send to you over Slack

berikv replied:
You need to set up Xcode and install the 15.4 simulator runtime to use this command. Let me know if that resolves the error.

amanjeetsingh150 commented on maestro-ios-xctest-runner/maestro-driver-iosUITests/Routes/Handlers/SessionInfoHandler.swift:
```diff
@@ -0,0 +1,33 @@
+import XCTest
+import Network
+
+@MainActor
```
Do we need to do this on main thread?

felipevolpone replied:
Yes, all other handlers also contain the `@MainActor`

berikv replied:
Thats not a great reason. The other handlers need @MainActor because they interact with XCTest.frawemork, which requires to be called on the main thread. I don't see code in the SessionInfoHandler that requires the main thread.

amanjeetsingh150 commented on maestro-android/src/androidTest/java/dev/mobile/maestro/MaestroDriverService.kt:
```diff
@@ -287,6 +287,14 @@ class Service(
+    ) {
+        responseObserver.onNext(sessionResponse { id = SdkService.sessionInfo.sessionId })
+        responseObserver.onCompleted()
+    }
```
I see changes here but not updated maestro server and maestro-android APKs

felipevolpone replied:
How can I do that?

felipevolpone replied:
Done

amanjeetsingh150 commented:
Can you add integration tests for this? I am finding hard to understand how and when users will write it in their flow file?

felipevolpone commented:
@amanjeetsingh150 I added an integration test, but I don't think it is that useful for this feature. I updated the PR description to explain more about this feature, let me know if it's not enough

felipevolpone commented on maestro-ios/src/main/java/ios/LocalIOSDevice.kt:
```diff
@@ -158,4 +159,8 @@ class LocalIOSDevice(
     }
+
+    override fun fetchSessionInfo(): Result<UUID, Throwable> {
+        return simctlIOSDevice.fetchSessionInfo()
```
Fixed 👍

berikv commented on maestro-ios-xctest-runner/maestro-driver-iosUITests/Routes/Handlers/SessionInfoHandler.swift:
```diff
@@ -0,0 +1,33 @@
+        category: String(describing: Self.self)
+    )
+    
+    private let sessionId = UUID()
```
Why is the XCTest runner involved to create a UUID? Seems like this can be done much faster and simpler in the worker..?

felipevolpone replied:
When the user uses the maestro SDK to use the Mock Server feature (https://maestro.mobile.dev/advanced/experimental/maestro-mock-server/getting-started#setup), it will proxy the app requests through our backend. The backend stores the events and then the maestro client needs to check if the data used in the `assertOutgoingRequest` matches the info stored.

The sessionId is related to a specific maestro session, meaning every time it runs `maestro test <flow>`. Because of that, the driver should create a new UUID every time and then, the client will get the sessionId and fetch the backend to get the events. That's why we can't use worker for that.

berikv replied:
Sounds like XCTest runner is really not the right place for that. It can be restarted throughout the flow and should not be used for storing UUID's. Can't maestro store the session UUID?

berikv replied:
How about the SDK doing a call-out to our backend with the sessionId to match on as a HTTP call?

axelniklasson replied:
@berikv the request to our backend with the sessionId to match on HTTP call happens from maestro and not from the SDK, as maestro needs the response in order to validate the assertions. Is there any way we can generate a session id from within the iOS SDK and keep that session id throughout the flow execution?

berikv replied:
What is a session exactly? How long does it live? Where does it depend on?

* If it is a flow run, then maestro should be the source of truth, eg Orchestra or SessionManager could keep this state.
* If a session is longer than a single flow run, then the source of truth needs to be outside of maestro (and not in the XCTest runner, which is part of maestro).
 * If a session is bound by the SDK, then let the SDK be the source of truth. If maestro needs to know the session ID, it can fetch it using HTTP calls from the backend or directly to the app, whatever works.
* If it is just a part of a flow run, the source of truth can still be maestro. A session object that's closed and reopened when needed. It could live alongside Orchestra.

amanjeetsingh150 commented:
> @amanjeetsingh150 I added an integration test, but I don't think it is that useful for this feature. I updated the PR description to explain more about this feature, let me know if it's not enough

Thanks @felipevolpone now I understand a bit more. Few questions:

1. Does these assertions can only be done with network mocking, right? If yes how do we plan to give a feedback to user if they try to use with live network?
2. I find it hard to understand where should I assert requests. Generally the requests are on user actions, and since requests can be in parallel I wouldn't which one to assert first?
3. Do we plan to include this assertion in flow file only?
